controllers/logicalvolume_controller.go

```diff
@@ -194,16 +195,51 @@ func (r *LogicalVolumeReconciler) createLV(ctx context.Context, log logr.Logger,
 			return nil
 		}
 
-		resp, err := r.lvService.CreateLV(ctx, &proto.CreateLVRequest{Name: string(lv.UID), DeviceClass: lv.Spec.DeviceClass, SizeGb: uint64(reqBytes >> 30)})
-		if err != nil {
-			code, message := extractFromError(err)
-			log.Error(err, message)
-			lv.Status.Code = code
-			lv.Status.Message = message
-			return err
+		var volume *proto.LogicalVolume
+
+		// Create a snapshot LV
+		if lv.Spec.Snapshot.DataSource != "" {
+			switch lv.Spec.Snapshot.AccessType {
+			case "ro", "rw":
+			default:
+				return fmt.Errorf("invalid access type for source volume: %s", lv.Spec.Snapshot.AccessType)
+
+			}
+			// Create a snapshot lv
+			resp, err := r.lvService.CreateLVSnapshot(ctx, &proto.CreateLVSnapshotRequest{
+				Name:         string(lv.UID),
+				DeviceClass:  lv.Spec.DeviceClass,
+				Sourcevolume: lv.Spec.Snapshot.DataSource,
+				SizeGb:       uint64(reqBytes >> 30),
+				AccessType:   lv.Spec.Snapshot.AccessType,
+				SnapType:     lv.Spec.Snapshot.Type,
+			})
+			if err != nil {
+				code, message := extractFromError(err)
+				log.Error(err, message)
+				lv.Status.Code = code
+				lv.Status.Message = message
+				return err
+			}
+			volume = resp.Snap
```

Comment:
This could be replaced by 
`lv.Status.VolumeID = resp.snap`
 or by changing the SnapshotResponse field name to volume .

Reply:
Since the volume can be a response of either a [snapshot request](https://github.com/topolvm/topolvm/pull/463/commits/d7ec57545a2b98f9fecc805d9c257cb6f7af1b7b#diff-7b9e88d2ad217bf9b768f9a41e9d4e857ce02ff9b66d0032138c573626d3459bR222) or a [volume request](https://github.com/topolvm/topolvm/pull/463/commits/d7ec57545a2b98f9fecc805d9c257cb6f7af1b7b#diff-7b9e88d2ad217bf9b768f9a41e9d4e857ce02ff9b66d0032138c573626d3459bR237); we fetch both of them as a volume and then update the value of `lv.Status.VolumeID` later [here](https://github.com/topolvm/topolvm/pull/463/commits/d7ec57545a2b98f9fecc805d9c257cb6f7af1b7b#diff-7b9e88d2ad217bf9b768f9a41e9d4e857ce02ff9b66d0032138c573626d3459bR240).